### ISSUE TITLE
budget assignment processed in parse order, not date order (BZ#1092)

### ISSUE THREAD

**Issue by tbm** [ledger/ledger#1092]
Note: the issue was created automatically with bugzilla2github

Bugzilla bug ID: BZ#1092
From: itsme

**Comment by helguson** [comment]
This seems to be the same problem as #111. There they say that it is not a bug but a feature.

**Comment by Qub1** [comment]
Seeing as hledger does provide this functionality by default, maybe it would be a good idea to include a flag that enables processing budget assignments in date order instead (or any other sorting order)?

Maybe something along the lines of `--balance-assignment-sort VEXPR`. That way users can sort them by date or index in the file if they wish.

I would be willing to write something up and submit a pull request if needed.

**Comment by jwiegley** [comment]
This is not a simple thing to add support for in C++ Ledger. Not impossible either, but it would require a fair bit of code reorganization. This is because processing of assertions and restrictions is done while parsing -- for speed -- rather than as a post-parsing phase.

**Comment by Qub1** [comment]
Yes, shortly after I wrote that comment I was looking through the code to see if maybe I could write it and submit a pull request but I came to the same conclusion as well. I was hoping it would be as simple as switching around sorting and processing but alas.

**Comment by tbm** [comment]
Can someone at least add something about this the docs? I don't think the limitation is currently documented.

**Comment by simonmichael** [comment]
I was trying to clean up/document hledger's implementation of this recently, which is mostly around [here](https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Data/Journal.hs#L623). As you can see the combination of inferring transaction-balancing amounts, balance assignment, balance assertions, arbitrary ordering, and posting dates, gets rather tricky. We have imposed a couple of restrictions to remove ambiguity. More restrictions could make this easier to implement.

**Comment by casr** [comment]
I just got tripped up by this as well. I put together this example to illustrate my use case.

`2018_19_general.ledger`:
```ledger
; Sales invoices
2019/09/01 ACME INDUSTRIES
    Assets:Accounts receivable                      1,000.00
    Liabilities:Tax payable                          -200.00
    Revenue                                          -800.00

; Cash
2019/09/30 Closing balances
    Assets:Accounts receivable                     -1,000.00 = 0
    Liabilities:Tax payable                           200.00 = 0
    Equity
```

` 2019_20_general.ledger`:
```ledger
; Sales invoices
2019/10/07 ACME INDUSTRIES
    Assets:Accounts receivable                      2,000.00
    Liabilities:Tax payable                          -400.00
    Revenue                                        -1,600.00

; Cash
2019/10/01 Opening balances
    ; Carried over from 2018/19
    Assets:Accounts receivable                      1,000.00 = 1,000.00
    Liabilities:Tax payable                           200.00 = 200.00
    Equity
```

I like to maintain my accounts split by year. I then post closing and opening balances so the files can be used on their own or combined together. I set aside a part of each year's accounts to collect invoices which means postings are out of date order. I use balance assertions to check accounts are at zero at the end of the year and to assert the correct amount is set at the opening balances.

However, when doing `ledger -f 2018_19_general.ledger -f 2019_20_general.ledger bal` you can expect to receive:
```
While parsing file "2019_20_general.ledger", line 10:
While parsing posting:
  Assets:Accounts receivable                      1,000.00 = 1,000.00
                                                             ^^^^^^^^
Error: Balance assertion off by -2000 (expected to see 2000)
```

**Comment by jwiegley** [comment]
This is a fundamental design decision at the moment, that needs to be revisited with a thorough design that takes into account both user expectations, performance, and correctness.
